src/v/storage/api.cc

```diff
@@ -52,4 +63,61 @@ void api::handle_disk_notification(
     }
 }
 
+ss::future<> api::monitor() {
+    vassert(ss::this_shard_id() == 0, "Run on wrong core");
+
+    /*
+     * The control loop should run periodically, but the current setting of 2
+     * seconds which provides decent responsiveness is too low to be used in
+     * practice long term. A follow up change will increase this setting and
+     * then allow cores to wake up this thread on demand to provide reactivity
+     * without frequent polling.
+     */
+    constexpr auto frequency = std::chrono::seconds(2);
+
+    while (true) {
+        try {
+            if (_log_storage_max_size().has_value()) {
+                co_await _monitor_sem.wait(
+                  frequency, std::max(_monitor_sem.current(), size_t(1)));
+            } else {
+                co_await _monitor_sem.wait();
+            }
+        } catch (const ss::semaphore_timed_out&) {
+            // time for some controlling
+        }
+
+        if (!_log_storage_max_size().has_value()) {
```

Comment:
It feels strange that this monitor loop is totally separate to the resource_mgmt one.

I was imagining that `disk_space_manager::run_loop` would use the usage information (it already calls `disk_usage()` to tell the raft (storage logs) layer its permitted disk space.

Reply:
> I was imagining that disk_space_manager::run_loop would use the usage information (it already calls disk_usage() to tell the raft (storage logs) layer its permitted disk space.

Yes, this is the plan. You can see how it is done for the cache here https://github.com/redpanda-data/redpanda/pull/11624

> It feels strange that this monitor loop is totally separate to the resource_mgmt one.

This is how I originally constructed it. But I thought the feedback would be something like: why isn't the `max_size` enforced in the storage layer the same way that the cache enforces its configured max size. So this PR is adding the storage capability of having a max size, rather than the logic for determining what the max size should be.

Does that align with your view?